elfo-core/src/message/repr.rs

```diff
@@ -0,0 +1,340 @@
+#[doc(hidden)]
+#[non_exhaustive] // must be created only via `MessageVTable::new()`
+pub struct MessageVTable {
+    pub(super) repr_layout: alloc::Layout, // of `MessageRepr<M>`
+    pub(super) name: &'static str,
+    pub(super) protocol: &'static str,
+    pub(super) labels: [Label; 2],    // protocol + name for `metrics`
+    pub(super) dumping_allowed: bool, // TODO: introduce `DumpingMode`.
+    #[cfg(feature = "network")]
+    pub(super) read_msgpack: unsafe fn(&[u8], NonNull<MessageRepr>) -> Result<(), decode::Error>,
+    #[cfg(feature = "network")]
+    #[allow(clippy::type_complexity)]
+    pub(super) write_msgpack:
+        unsafe fn(NonNull<MessageRepr>, &mut Vec<u8>, usize) -> Result<(), encode::Error>,
+    pub(super) debug: unsafe fn(NonNull<MessageRepr>, &mut fmt::Formatter<'_>) -> fmt::Result,
+    pub(super) clone: unsafe fn(NonNull<MessageRepr>, NonNull<MessageRepr>),
+    // TODO: remove and use `as_serialize_any` in the dumper after benchmarking.
+    pub(super) erase: unsafe fn(NonNull<MessageRepr>) -> dumping::ErasedMessage,
+    pub(super) as_serialize_any:
+        unsafe fn(NonNull<MessageRepr>) -> NonNull<dyn erased_serde::Serialize>,
+    pub(super) deserialize_any: unsafe fn(
+        &mut dyn erased_serde::Deserializer<'_>,
+        NonNull<MessageRepr>,
+    ) -> Result<(), erased_serde::Error>,
+    pub(super) drop: unsafe fn(NonNull<MessageRepr>),
+}
+
+impl MessageVTable {
+    /// Creates a new vtable for the provided message type.
+    /// This is the only way to create a vtable.
+    // Reexported in `elfo::_priv`.
+    #[doc(hidden)]
+    pub const fn new<M: Message>(
+        name: &'static str,
+        protocol: &'static str,
+        dumping_allowed: bool,
+    ) -> Self {
+        Self {
+            repr_layout: alloc::Layout::new::<MessageRepr<M>>(),
+            name,
+            protocol,
+            labels: [
+                Label::from_static_parts("message", name),
+                Label::from_static_parts("protocol", protocol),
+            ],
+            dumping_allowed,
+            debug: vtablefns::debug::<M>,
+            clone: vtablefns::clone::<M>,
+            erase: vtablefns::erase::<M>,
+            as_serialize_any: vtablefns::as_serialize_any::<M>,
+            deserialize_any: vtablefns::deserialize_any::<M>,
+            drop: vtablefns::drop::<M>,
+            #[cfg(feature = "network")]
+            read_msgpack: vtablefns::read_msgpack::<M>,
+            #[cfg(feature = "network")]
+            write_msgpack: vtablefns::write_msgpack::<M>,
+        }
+    }
+}
+
+/// Generic vtable's functions for monomorphization in [`MessageVTable::new()`].
+///
+/// All functions are `unsafe` because they work with raw pointers.
+///
+/// # Safety
+///
+/// Common safety requirements for all functions:
+/// * input pointers (`ptr`) must be [valid] for reading `MessageRepr<M>`.
+/// * output pointers (`out_ptr`) must be [valid] for writing `MessageRepr<M>`.
```

Comment:
Output pointers also should not contain any previous values, given that `ptr::write()` will discard them without running destructors.

Reply:
I wonder if maybe having a separate type for an “uninit `MessageRepr`” would make more sense here (so `alloc_repr()` would return `UninitMessageRepr`, and e.g. `clone()` would accept one as `out_ptr`).

Reply:
TLDR: It's nice idea, but leads to overcomplicated code